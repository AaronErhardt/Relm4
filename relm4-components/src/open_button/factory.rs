use gtk::prelude::ButtonExt;
use relm4::factory::{DynamicIndex, FactoryPrototype, FactoryVecDeque};
use relm4::{gtk, send, WidgetPlus};

use super::OpenButtonMsg;

use std::path::PathBuf;

#[derive(Debug)]
pub(crate) struct FileListItem {
    pub(crate) path: PathBuf,
}

#[derive(Debug)]
pub(crate) struct FileListItemWidgets {
    label: gtk::Button,
    row: gtk::ListBoxRow,
}

impl FactoryPrototype for FileListItem {
    type Factory = FactoryVecDeque<Self>;
    type View = gtk::Box;
    type Widgets = FileListItemWidgets;
    type Root = gtk::ListBoxRow;
    type Msg = OpenButtonMsg;

    fn init_view(&self, key: &DynamicIndex, sender: relm4::Sender<Self::Msg>) -> Self::Widgets {
        let label = gtk::Button::with_label(
            self.path
                .iter()
                .last()
                .expect("Empty path")
                .to_str()
                .expect("Couldn't convert path to string"),
        );
        let row = gtk::ListBoxRow::builder().child(&label).build();

        label.inline_css(b"margin: 0;");

        let key = key.clone();
        label.connect_clicked(move |_| {
            send!(sender, OpenButtonMsg::OpenRecent(key.clone()));
        });
        FileListItemWidgets { label, row }
    }

    fn view(&self, _key: &DynamicIndex, widgets: &Self::Widgets) {
        widgets.label.set_label(
            self.path
                .iter()
                .last()
                .expect("Empty path")
                .to_str()
                .expect("Couldn't convert path to string"),
        );
    }

    fn position(&self, _key: &DynamicIndex) {}

    fn root_widget(widgets: &Self::Widgets) -> &Self::Root {
        &widgets.row
    }
}
