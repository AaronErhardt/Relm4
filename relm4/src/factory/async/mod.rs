mod builder;
mod collections;
mod component_storage;
mod future_data;
mod handle;
mod traits;

pub use builder::AsyncFactoryBuilder;
use future_data::AsyncData;
pub use handle::AsyncFactoryHandle;

pub use collections::{AsyncFactoryVecDeque, AsyncFactoryVecDequeGuard};
pub use component_storage::AsyncComponentStorage;
pub use traits::AsyncFactoryComponent;
