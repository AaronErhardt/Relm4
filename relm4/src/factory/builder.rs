use super::{handle::FactoryHandle, DynamicIndex, FactoryComponent, FactoryView};

use crate::{shutdown, Receiver, Sender};

use std::cell::RefCell;
use std::rc::Rc;

use async_oneshot::oneshot;
use futures::FutureExt;

#[allow(missing_debug_implementations)]
pub(super) struct FactoryBuilder<Widget, C, ParentMsg>
where
    Widget: FactoryView,
    C: FactoryComponent<Widget, ParentMsg>,
    ParentMsg: 'static,
{
    pub(super) data: Rc<RefCell<C>>,
    pub(super) root_widget: C::Root,
    pub(super) input_tx: Sender<C::Input>,
    pub(super) input_rx: Receiver<C::Input>,
    pub(super) output_tx: Sender<C::Output>,
    pub(super) output_rx: Receiver<C::Output>,
}

impl<Widget, C, ParentMsg> FactoryBuilder<Widget, C, ParentMsg>
where
    Widget: FactoryView,
    C: FactoryComponent<Widget, ParentMsg>,
    ParentMsg: 'static,
{
    pub(super) fn new(index: &DynamicIndex, params: C::InitParams) -> Self {
        // Used for all events to be processed by this component's internal service.
        let (input_tx, input_rx) = crate::channel::<C::Input>();

        // Used by this component to send events to be handled externally by the caller.
        let (output_tx, output_rx) = crate::channel::<C::Output>();

        let data = Rc::new(RefCell::new(C::init_model(
            params, index, &input_tx, &output_tx,
        )));

        let root_widget = C::init_root();

        Self {
            data,
            root_widget,
            input_tx,
            input_rx,
            output_tx,
            output_rx,
        }
    }

    /// Starts the component, passing ownership to a future attached to a GLib context.
    pub(super) fn launch<Transform>(
        self,
        index: &DynamicIndex,
        returned_widget: Widget::ReturnedWidget,
        parent_sender: &Sender<ParentMsg>,
        transform: Transform,
    ) -> FactoryHandle<Widget, C, ParentMsg>
    where
        Transform: Fn(C::Output) -> Option<ParentMsg> + 'static,
    {
        let Self {
            data,
            root_widget,
            input_tx,
            mut input_rx,
            output_tx,
            mut output_rx,
        } = self;

        let forward_sender = parent_sender.0.clone();
        crate::spawn_local(async move {
            while let Some(msg) = output_rx.recv().await {
                if let Some(new_msg) = transform(msg) {
                    if forward_sender.send(new_msg).is_err() {
                        break;
                    }
                }
            }
        });

        // Sends messages from commands executed from the background.
        let (cmd_tx, mut cmd_rx) = crate::channel::<C::CommandOutput>();

        // Gets notifications when a component's model and view is updated externally.
        let (notifier, notifier_rx) = flume::bounded(0);

        let mut widgets = data.borrow_mut().init_widgets(
            index,
            &root_widget,
            &returned_widget,
            &input_tx,
            &output_tx,
        );

        // The source ID of the component's service will be sent through this once the root
        // widget has been iced, which will give the component one last chance to say goodbye.
        let (burn_notifier, burn_recipient) = oneshot::<gtk::glib::SourceId>();

        // Notifies the component's child commands that it is now deceased.
        let (death_notifier, death_recipient) = shutdown::channel();

        let input_tx_ = input_tx.clone();
        let runtime_data = data.clone();

        // Spawns the component's service. It will receive both `Self::Input` and
        // `Self::CommandOutput` messages. It will spawn commands as requested by
        // updates, and send `Self::Output` messages externally.
        let id = crate::spawn_local(async move {
            let mut burn_notice = burn_recipient.fuse();
            loop {
                let notifier = notifier_rx.recv_async().fuse();
                let cmd = cmd_rx.recv().fuse();
                let input = input_rx.recv().fuse();

                futures::pin_mut!(cmd);
                futures::pin_mut!(input);
                futures::pin_mut!(notifier);

                let _ = futures::select!(
                    // Performs the model update, checking if the update requested a command.
                    // Runs that command asynchronously in the background using tokio.
                    message = input => {
                        if let Some(message) = message {
                            let mut model = runtime_data.borrow_mut();

                            if let Some(command) = model.update_with_view(&mut widgets, message, &input_tx_, &output_tx)
                            {
                                let recipient = death_recipient.clone();
                                crate::spawn(C::command(command, recipient, cmd_tx.clone()));
                            }
                        }
                    }

                    // Handles responses from a command.
                    message = cmd => {
                        if let Some(message) = message {
                            let mut model = runtime_data.borrow_mut();
                            model.update_cmd_with_view(&mut widgets, message, &input_tx_, &output_tx);
                        }
                    }

                    // Triggered when the model and view have been updated externally.
                    _ = notifier => {
                        let model = runtime_data.borrow_mut();
                        model.update_view(&mut widgets, &input_tx_, &output_tx);
                    }

                    // Triggered when the component is destroyed
                    id = burn_notice => {
                        let mut model = runtime_data.borrow_mut();

                        model.shutdown(&mut widgets, output_tx);

                        death_notifier.shutdown();

                        if let Ok(id) = id {
                            id.remove();
                        }

                        return
                    }
                );
            }
        });

        // Give back a type for controlling the component service.
        FactoryHandle {
            data,
            root_widget,
            returned_widget,
            input: input_tx,
            notifier: Sender(notifier),
            burner: crate::component::CompBurner {
                runtime_id: RefCell::new(Some(id)),
                burn_notifier,
            },
        }
    }
}
