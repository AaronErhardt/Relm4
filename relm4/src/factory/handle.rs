use super::FactoryComponent;

use std::cell::RefCell;
use std::rc::Rc;

use super::FactoryView;
use crate::Sender;

#[derive(Debug)]
pub(super) struct FactoryHandle<Widget, C: FactoryComponent<Widget, ParentMsg>, ParentMsg>
where
    Widget: FactoryView,
    C: FactoryComponent<Widget, ParentMsg>,
{
    pub(super) data: Rc<RefCell<C>>,
    pub(super) root_widget: C::Root,
    pub(super) returned_widget: Widget::ReturnedWidget,
    pub(super) input: Sender<C::Input>,
    pub(super) notifier: Sender<()>,

    /// Kills the event loop of this component on drop.
    pub(super) burner: crate::component::CompBurner,
}
