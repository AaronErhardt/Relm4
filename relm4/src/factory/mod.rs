//! Defines traits and data types to generate widgets from collections efficiently.

mod dynamic_index;

/// Traits and implementations used for factories to interact with widgets.
pub mod widgets;

/// Implementation of asynchronous factories.
pub mod r#async;
pub mod positions;

/// Implementation of regular factories.
pub mod sync;

mod data_guard;
use data_guard::DataGuard;

pub use r#async::{AsyncFactoryComponent, AsyncFactoryVecDeque, AsyncFactoryVecDequeGuard};
pub use sync::{
    CloneableFactoryComponent, FactoryComponent, FactoryVecDeque, FactoryVecDequeGuard,
};

pub use crate::channel::{AsyncFactorySender, FactorySender};
pub use dynamic_index::DynamicIndex;
pub use widgets::traits::*;
