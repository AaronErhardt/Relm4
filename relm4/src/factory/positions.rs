//! Position types for various widgets.

/// Storing information about where new widgets can be placed
/// inside a [`gtk::Grid`].
#[derive(Debug)]
pub struct GridPosition {
    /// The number of the column.
    pub column: i32,
    /// The number of the row.
    pub row: i32,
    /// The amount of columns the widget should take.
    pub width: i32,
    /// The amount of rows the widget should take.
    pub height: i32,
}

#[derive(Debug)]
/// Position used for [`gtk::Fixed`].
pub struct FixedPosition {
    /// Position on the x-axis.
    pub x: f64,
    /// Position on the y-axis.
    pub y: f64,
}

#[derive(Debug, Default)]
/// Position (or rather information) for a stack page
/// inside a factory.
pub struct StackPageInfo {
    /// The name of the [`gtk::StackPage`].
    pub name: Option<String>,
    /// The title of the [`gtk::StackPage`].
    pub title: Option<String>,
}
