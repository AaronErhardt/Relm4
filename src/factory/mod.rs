//! Defines traits and data types used to efficiently generating widgets from collections.

use gtk::glib::Sender;
use gtk::prelude::WidgetExt;

pub mod collections;
pub mod positions;
mod widgets;

pub use collections::*;

/// Define behavior to create, update and remove widgets according to
/// data stored in a factory.
pub trait FactoryPrototype: Sized {
    /// Factory container that stores the data.
    type Factory: Factory<Self, Self::View>;

    /// Type that stores all widgets needed to update them
    /// in the [`view`](FactoryPrototype::view) function.
    type Widgets: std::fmt::Debug;

    /// Outermost type of the newly created widgets.
    /// Similar to the `Root` type in [`crate::Widgets`].
    type Root: WidgetExt;

    /// Widget that the generated widgets are added to.
    type View: FactoryView<Self::Root>;

    /// Message type used to send messages back to the component or app
    /// where this factory is used
    type Msg;

    /// Create new widgets when self is inserted into the factory.
    fn init_view(
        &self,
        key: &<Self::Factory as Factory<Self, Self::View>>::Key,
        sender: Sender<Self::Msg>,
    ) -> Self::Widgets;

    /// Set the widget position upon creation, useful for [`gtk::Grid`] or similar.
    fn position(
        &self,
        key: &<Self::Factory as Factory<Self, Self::View>>::Key,
    ) -> <Self::View as FactoryView<Self::Root>>::Position;

    /// Function called when self is modified.
    fn view(
        &self,
        key: &<Self::Factory as Factory<Self, Self::View>>::Key,
        widgets: &Self::Widgets,
    );

    /// Get the outermost widget from the widgets.
    fn root_widget(widgets: &Self::Widgets) -> &Self::Root;
}

/// A container that is a able to efficiently update, generate and remove widgets
/// that represent the data stored in the container.
pub trait Factory<Data, View>
where
    Data: FactoryPrototype<View = View>,
    View: FactoryView<Data::Root>,
{
    /// Key that provides additional information for the [`FactoryPrototype`] functions.
    type Key: ?Sized;

    /// Efficiently update the view according to data changes.
    fn generate(&self, view: &View, sender: Sender<Data::Msg>);
}

/*/// A trait to simplify the implementation of [`FactoryView`] for most
/// GTK4 widgets.
trait SimpleFactoryView<Widget: std::fmt::Debug> {
    /// Position type used by this widget.
    ///
    /// For example [`GridPosition`] for [`gtk::Grid`] or `()` for [`gtk::Box`]
    type Position;

    /// Adds a new widget to self at the end.
    fn add(&self, widget: &Widget, position: &Self::Position);

    /// Removes a widget from self at the end.
    fn remove(&self, widget: &Widget);
}*/

/// A trait implemented for GTK4 widgets that allows a factory to create and remove widgets.
pub trait FactoryView<Widget> {
    /// Widget type that's stored inside a factory data type.
    type Root: std::fmt::Debug;

    /// Position type used by this widget.
    ///
    /// For example [`GridPosition`](positions::GridPosition) for [`gtk::Grid`] or `()` for [`gtk::Box`]
    type Position;

    /// Adds a new widget to self at the end.
    fn add(&self, widget: &Widget, position: &Self::Position) -> Self::Root;

    /// Removes a widget from self at the end.
    fn remove(&self, widget: &Self::Root);
}

/*impl<Widget, M> FactoryView<Widget> for M
where
    M: SimpleFactoryView<Widget>,
    Widget: Clone + std::fmt::Debug,
{
    type Position = M::Position;
    type Root = Widget;

    fn add(&self, widget: &Widget, position: &Self::Position) -> Self::Root {
        self.add(widget, position);
        widget.clone()
    }

    fn remove(&self, widget: &Widget) {
        self.remove(widget);
    }
}*/

/// Extends [`FactoryView`] for containers that work similar to lists.
/// This means that the container can insert widgets before and after other
/// widgets.
pub trait FactoryListView<Widget>
where
    Self: FactoryView<Widget>,
{
    /// Insert a widget after another widget.
    fn insert_after(&self, widget: &Widget, other: &Self::Root) -> Self::Root;

    /// Add an widget to the front.
    fn push_front(&self, widget: &Widget) -> Self::Root;
}
