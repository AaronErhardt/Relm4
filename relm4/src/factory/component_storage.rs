use crate::factory::{DynamicIndex, FactoryBuilder, FactoryComponent, FactoryHandle, FactoryView};
use crate::Sender;

use std::cell::{Ref, RefMut};
use std::rc::Rc;

#[allow(missing_debug_implementations)]
pub(super) enum ComponentStorage<Widget, C, ParentMsg>
where
    Widget: FactoryView,
    C: FactoryComponent<Widget, ParentMsg>,
    ParentMsg: 'static,
    Widget: 'static,
{
    Builder(FactoryBuilder<Widget, C, ParentMsg>),
    Final(FactoryHandle<Widget, C, ParentMsg>),
}

impl<Widget, C, ParentMsg> ComponentStorage<Widget, C, ParentMsg>
where
    Widget: FactoryView,
    C: FactoryComponent<Widget, ParentMsg>,
{
    pub(super) fn get(&self) -> Ref<'_, C> {
        match self {
            Self::Builder(builder) => builder.data.borrow(),
            Self::Final(handle) => handle.data.borrow(),
        }
    }

    pub(super) fn get_mut(&mut self) -> RefMut<'_, C> {
        match self {
            Self::Builder(builder) => builder.data.borrow_mut(),
            Self::Final(handle) => handle.data.borrow_mut(),
        }
    }

    pub(super) fn widget(&self) -> &C::Root {
        match self {
            Self::Builder(builder) => &builder.root_widget,
            Self::Final(handle) => &handle.root_widget,
        }
    }

    pub(super) fn send(&self, msg: C::Input) {
        match self {
            Self::Builder(builder) => builder.input_tx.send(msg),
            Self::Final(handle) => handle.input.send(msg),
        }
    }

    pub(super) fn state_change_notify(&self) {
        if let Self::Final(handle) = self {
            handle.notifier.send(());
        }
    }

    pub(super) fn extract(self) -> C {
        match self {
            Self::Builder(builder) => Rc::try_unwrap(builder.data).unwrap().into_inner(),
            Self::Final(handle) => {
                if let Some(id) = handle.burner.runtime_id.borrow_mut().take() {
                    id.remove();
                }
                Rc::try_unwrap(handle.data).unwrap().into_inner()
            }
        }
    }

    pub(super) fn launch(
        self,
        index: &DynamicIndex,
        returned_widget: Widget::ReturnedWidget,
        parent_sender: &Sender<ParentMsg>,
    ) -> Option<Self> {
        if let Self::Builder(builder) = self {
            Some(Self::Final(builder.launch(
                index,
                returned_widget,
                parent_sender,
                C::output_to_parent_msg,
            )))
        } else {
            None
        }
    }

    pub(super) fn returned_widget(&self) -> Option<&Widget::ReturnedWidget> {
        if let Self::Final(handle) = self {
            Some(&handle.returned_widget)
        } else {
            None
        }
    }
}
