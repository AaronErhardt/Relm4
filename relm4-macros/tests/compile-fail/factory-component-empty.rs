use relm4::factory::FactoryComponent;

#[derive(Debug)]
struct TestFactoryComponent;

#[relm4_macros::factory]
impl FactoryComponent for TestFactoryComponent {

}

fn main() {}
