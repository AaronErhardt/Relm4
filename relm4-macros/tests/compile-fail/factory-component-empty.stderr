error: expected `view!` macro invocation
 --> tests/compile-fail/factory-component-empty.rs:7:1
  |
7 | / impl FactoryComponent for TestFactoryComponent {
8 | |
9 | | }
  | |_^

error[E0046]: not all trait items implemented, missing: `ParentWidget`, `ParentInput`, `CommandOutput`, `Input`, `Output`, `Init`, `Root`, `init_model`, `init_root`, `init_widgets`
 --> tests/compile-fail/factory-component-empty.rs:7:1
  |
7 | impl FactoryComponent for TestFactoryComponent {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ missing `ParentWidget`, `ParentInput`, `CommandOutput`, `Input`, `Output`, `Init`, `Root`, `init_model`, `init_root`, `init_widgets` in implementation
  |
  = help: implement the missing item: `type ParentWidget = Type;`
  = help: implement the missing item: `type ParentInput = Type;`
  = help: implement the missing item: `type CommandOutput = Type;`
  = help: implement the missing item: `type Input = Type;`
  = help: implement the missing item: `type Output = Type;`
  = help: implement the missing item: `type Init = Type;`
  = help: implement the missing item: `type Root = Type;`
  = help: implement the missing item: `fn init_model(_: <Self as FactoryComponent>::Init, _: &DynamicIndex, _: FactoryComponentSender<Self>) -> Self { todo!() }`
  = help: implement the missing item: `fn init_root(&self) -> <Self as FactoryComponent>::Root { todo!() }`
  = help: implement the missing item: `fn init_widgets(&mut self, _: &DynamicIndex, _: &<Self as FactoryComponent>::Root, _: &<<Self as FactoryComponent>::ParentWidget as FactoryView>::ReturnedWidget, _: FactoryComponentSender<Self>) -> <Self as FactoryComponent>::Widgets { todo!() }`
